src/ray/object_manager/object_manager.cc

```diff
@@ -332,57 +332,60 @@ void ObjectManager::HandleReceiveFinished(const ObjectID &object_id,
+    PushInternal(object_id, node_id, absl::optional<std::string>());
     return;
   }
+  auto object_url = get_spilled_object_url_(object_id);
+  if (object_url.empty()) {
+    RAY_LOG(WARNING) << "Attempting to push object " << object_id
+                     << " which is not local. It may have been evicted.";
+    return;
+  }
+  // Push the object directly from spilled object, bypassing
+  // local store restoration.
+  PushInternal(object_id, node_id, absl::optional<std::string>(object_url));
+}
 
+void ObjectManager::PushInternal(const ObjectID &object_id, const NodeID &node_id,
+                                 absl::optional<std::string> spilled_url) {
   auto rpc_client = GetRpcClient(node_id);
   if (rpc_client) {
-    const ObjectInfo &object_info = local_objects_[object_id].object_info;
-    uint64_t data_size =
-        static_cast<uint64_t>(object_info.data_size + object_info.metadata_size);
-    uint64_t metadata_size = static_cast<uint64_t>(object_info.metadata_size);
-    uint64_t num_chunks = buffer_pool_.GetNumChunks(data_size);
+    std::shared_ptr<SpilledObject> spilled_object;
+    if (spilled_url.has_value()) {
+      auto optional_spilled_object = SpilledObject::CreateSpilledObject(
+          spilled_url.value(), config_.object_chunk_size);
+      if (!optional_spilled_object.has_value()) {
+        RAY_LOG(WARNING) << "Failed to load splled object " << object_id
+                         << ". It may have been evicted.";
+        return;
+      }
+      spilled_object =
+          std::make_shared<SpilledObject>(std::move(optional_spilled_object.value()));
+    }
 
+    uint64_t data_size;
+    uint64_t metadata_size;
+    uint64_t num_chunks;
     rpc::Address owner_address;
-    owner_address.set_raylet_id(object_info.owner_raylet_id.Binary());
-    owner_address.set_ip_address(object_info.owner_ip_address);
-    owner_address.set_port(object_info.owner_port);
-    owner_address.set_worker_id(object_info.owner_worker_id.Binary());
+
+    if (spilled_object) {
+      data_size = spilled_object->GetDataSize() + spilled_object->GetMetadataSize();
+      metadata_size = spilled_object->GetMetadataSize();
+      num_chunks = spilled_object->GetNumChunks();
+      owner_address = spilled_object->GetOwnerAddress();
+    } else {
+      const ObjectInfo &object_info = local_objects_[object_id].object_info;
+      data_size =
+          static_cast<uint64_t>(object_info.data_size + object_info.metadata_size);
+      metadata_size = static_cast<uint64_t>(object_info.metadata_size);
+      num_chunks = buffer_pool_.GetNumChunks(data_size);
+
+      owner_address.set_raylet_id(object_info.owner_raylet_id.Binary());
+      owner_address.set_ip_address(object_info.owner_ip_address);
+      owner_address.set_port(object_info.owner_port);
+      owner_address.set_worker_id(object_info.owner_worker_id.Binary());
+    }
 
     RAY_LOG(DEBUG) << "Sending object chunks of " << object_id << " to node " << node_id
```

Comment:
Is it possible to have `PushInternal`/`SendObjectChunk` take in a generic memory buffer, so we don't have to treat spilled objects differently from plasma objects? Just wondering because I think it could make this code more robust (don't have to do a bunch of if branches).

Reply:
I do see a viable way forward is to create an interface like ObjectForPush, where we have SpilledObject implementation and BufferedPool implementation.
Other alternatives, such as hiding spilled object underneath ObjectBufferPool might require a huge amount of refactor. In particular the information (data, and owner address)  needed for pushing a plasma objected are spread out  in both ObjectBufferPool and ObjectManager.local_objects_.
Let me know your thought!

Reply:
We'll have offline discussion today.

Reply:
Conclusion: Merge the current PR without clean up -> Have ObjectForPush abstraction in the follow-up PR